src/Microsoft.AspNetCore.Blazor.Browser.JS/src/Services/Http.ts

```diff
@@ -1,50 +1,76 @@
 const httpClientTypeName = 'BrowserHttpMessageHandler';
 const httpClientFullTypeName = `${httpClientNamespace}.${httpClientTypeName}`;
 let receiveResponseMethod: MethodHandle;
+let allocateArrayMethod: MethodHandle;
 
-registerFunction(`${httpClientFullTypeName}.Send`, (id: number, method: string, requestUri: string, body: string | null, headersJson: string | null, fetchArgs: RequestInit | null) => {
-  sendAsync(id, method, requestUri, body, headersJson, fetchArgs);
+registerFunction(`${httpClientFullTypeName}.Send`, (id: number, body: System_Array<any>, jsonFetchArgs: System_String) => {
+  sendAsync(id, body, jsonFetchArgs);
 });
 
-async function sendAsync(id: number, method: string, requestUri: string, body: string | null, headersJson: string | null, fetchArgs: RequestInit | null) {
+async function sendAsync(id: number, body: System_Array<any>, jsonFetchArgs: System_String) {
   let response: Response;
-  let responseText: string;
+  let responseData: ArrayBuffer;
 
-  const requestInit: RequestInit = fetchArgs || {};
-  requestInit.method = method;
-  requestInit.body = body || undefined;
+  const dotNetRequestInit: DotNetRequestInit = JSON.parse(platform.toJavaScriptString(jsonFetchArgs));
+  const requestInit: RequestInit = {};
 
-  try {
-    requestInit.headers = headersJson ? (JSON.parse(headersJson) as string[][]) : undefined;
+  // discard nulls and requestUri
+  for (var key in dotNetRequestInit) {
+    if (key != "requestUri" && dotNetRequestInit.hasOwnProperty(key) && dotNetRequestInit[key]) {
+      requestInit[key] = dotNetRequestInit[key];
+    }
+  }
 
-    response = await fetch(requestUri, requestInit);
-    responseText = await response.text();
+  requestInit.body = body ? platform.getByteArray(body) : undefined;
+
+  try {
+    response = await fetch(dotNetRequestInit.requestUri, requestInit);
+    responseData = await response.arrayBuffer();
   } catch (ex) {
     dispatchErrorResponse(id, ex.toString());
     return;
   }
 
-  dispatchSuccessResponse(id, response, responseText);
+  dispatchSuccessResponse(id, response, responseData);
 }
 
-function dispatchSuccessResponse(id: number, response: Response, responseText: string) {
+function dispatchSuccessResponse(id: number, response: Response, responseData: ArrayBuffer) {
   const responseDescriptor: ResponseDescriptor = {
     StatusCode: response.status,
+    StatusText: response.statusText,
     Headers: []
   };
   response.headers.forEach((value, name) => {
     responseDescriptor.Headers.push([name, value]);
   });
 
+  if (!allocateArrayMethod) {
+    allocateArrayMethod = platform.findMethod(
+      httpClientAssembly,
+      httpClientNamespace,
+      httpClientTypeName,
+      'AllocateArray'
+    );
+  }
+
+  // allocate a managed byte[] of the right size
+  const dotNetArray = platform.callMethod(allocateArrayMethod, null, [platform.toDotNetString(responseData.byteLength.toString())]) as System_Array<any>;
```

Comment:
Did you have to convert the length to a `string` because of a JS interop limitation? Fine if so. Just wondered whether you tried passing it directly as a number and receiving an `int` on the .NET side. Not sure whether that works yet.

Reply:
I've copied the trick from https://github.com/aspnet/Blazor/blob/dc532a3963a36597c2d4df907c9a686fd58266b7/src/Microsoft.AspNetCore.Blazor.Browser.JS/src/Services/Http.ts#L71-L76
so it's prior art.

callMethod expects an `System_Object[]` but i suppose it would work in theory as it seems to allocates 32bit space on the stack for each argument, which seems a bit fishy https://github.com/aspnet/Blazor/blob/42e3848837be69a479944e55a8789ecc59a4f2cc/src/Microsoft.AspNetCore.Blazor.Browser.JS/src/Platform/Mono/MonoPlatform.ts#L50-L64
Everything around `RenderTree*` friends is built with 64bit pointers in mind.

Reply:
All the .NET code has to allow for running in x64 (because that will often be the case, e.g., in SSR or unit tests). But the JS-side code can assume 32-bit because the only time JS code can see the .NET memory directly is when running under WebAssembly, and the WebAssembly runtime is always 32-bit. Perhaps this will change in the future but it is the case presently.